mainchain-follower/db-sync-follower/src/candidates/mod.rs

```diff
@@ -133,19 +133,38 @@ impl CandidateDataSource for CandidatesDataSourceImpl {
 		Ok(AriadneParameters { d_parameter: d_param, permissioned_candidates: candidates })
 	}
 
+	// Get all registered candidate registrations for the given epoch.
+	// Note, this method requests only ADA candidates from Cardano.
+	// TODO ETH: From Minotaur perspective, there is a choice to make, whether CandidateDataSource
+	// is responsible to request only ADA candidates or ETH candidates as well.
 	async fn get_candidates(
 			&self,
 			epoch: McEpochNumber,
 			committee_candidate_address: MainchainAddress
 		)-> Result<Vec<CandidateRegistrations>> {
 		let epoch = EpochNumber::from(self.get_epoch_of_data_storage(epoch)?);
-		let candidates = self.get_registered_candidates(epoch, committee_candidate_address).await?;
+
+		// get registered candidates for ADA
+		let ada_candidates = self
+			.get_registered_candidates(epoch, committee_candidate_address)
+			.await?;
+
+		// TODO ETH: get registered candidates for ETH if CandidateDataSource is acting as unified facade
+
 		let stake_map = Self::make_stake_map(db_model::get_stake_distribution(&self.pool, epoch).await?);
-		Ok(Self::group_candidates_by_mc_pub_key(candidates).into_iter().map(|(mainchain_pub_key, candidate_registrations)| {
+		Ok(Self::group_candidates_by_mc_pub_key(ada_candidates).into_iter().map(|(mainchain_pub_key, candidate_registrations)| {
 			CandidateRegistrations {
 				mainchain_pub_key: mainchain_pub_key.clone(),
-				registrations: candidate_registrations.into_iter().map(Self::make_registration_data).collect(),
-				stake_delegation: Self::get_stake_delegation(&stake_map, &mainchain_pub_key),
+				eth_pub_key: None,  // TODO ETH: add ETH pub key
+				registrations: Registrations {
+						ada_registrations: candidate_registrations
+							.into_iter()
+							.map(Self::make_registration_data)
+							.collect(),
+						eth_registrations: vec![]  // TODO ETH: add ETH registrations
```

Comment:
Eth registrations will be observed on Ethereum, won't they? If so, I'd rather have a separate module for following Ethereum. This way, the code will be more modular and it will be easier for the users to pick up only the parts they are interested in.

Reply:
Yes, observed on Ethereum, but I would suggest to have 2 levels of data sources:
- level 1: data source which is passed everywhere (which is currently CandidateDataSource)
- level 2: one resource-specific data source for each resource type (CardanoDS, EthereumDS, etc.)

Thus defined CandidateDataSource will aggregate information from different sources.
From this POV, the next step would be to define a ResourceDS trait and move the Cardano-specific code there from CandidateDataSource.

Reply:
What you wrote looks fine. However,  I would suggest a different approach on getting there. Since this code already is a CardanoDS, instead of modifying it, create a separate module (or even a simple function in authority-selection-inherents for now) which will glue the results of CardanoDS and EthereumDS. This way the diff will be easier to read (especially of the next PR) and the change will be done in less invasive way.